Document TCP RTO vs QUIC PTO (#3441)
quicwg/base-drafts: draft-ietf-quic-recovery.md

when the corresponding acknowledgment is sent, allowing a peer to maintain a
more accurate round-trip time estimate (see Section 13.2 of {{QUIC-TRANSPORT}}).

### Probe Timeout Replaces RTO and TLP

QUIC uses a probe timeout (see {{pto}}), with a timer based on TCP's

> RTO timeout.  QUIC PTO includes the peer's max_ack_delay in the calculation,

Reviewer: ```suggestion
RTO period computation.  QUIC's PTO period includes the peer's maximum expected acknowledgement delay (max_ack_delay, Section XX),
```

> instead of relying upon a fixed minimum RTO.  Unlike TCP's RTO, which collapses

Reviewer: ```suggestion
instead of using a fixed minimum period.  Unlike TCP, which collapses
```

> the congestion window upon expiry, QUIC does not change the congestion window

Reviewer: ```suggestion
the congestion window upon expiry of an RTO, QUIC does not change the congestion window
```

> and instead allows sending probe packets whenever the timer expires.  This is

Reviewer: ```suggestion
and instead sends probe packets whenever the timer expires.  This is
```

similar to TCP with F-RTO, but it does allow more packets to be sent when the

> congestion window was not fully utilized when the probe timeout expires. Though

Reviewer: I don't think this is similar to F-RTO... I would remove this previous sentence.

Reviewer: ping

Author: Why not?

this is slightly more aggressive than TCP RTO, it's less aggressive than the

> congestion controller allowed.

Reviewer 2: Phrasing here suggests "than the congestion controller allows" or "than the congestion controller in TCP allowed" or something else.

Author: Updated to compare to a non-app limited connection.  Tell me if that works.

> connection was not application limited.

Reviewer: I'm not sure I understand this last sentence... how is app-limited relevant here?

Reviewer: ping

Author: If you weren't app-limited, then you wouldn't have CWND free to send more packets after the PTO expires.

A single packet loss at the tail does not indicate persistent congestion, so

> QUIC defines a time-based definition (see {{persistent-congestion}}) to

Reviewer: ```suggestion
QUIC specifies a time-based definition (see {{persistent-congestion}}) to
```

when the corresponding acknowledgment is sent, allowing a peer to maintain a
more accurate round-trip time estimate (see Section 13.2 of {{QUIC-TRANSPORT}}).

### Probe Timeout Replaces RTO and TLP

QUIC uses a probe timeout (see {{pto}}), with a timer based on TCP's RTO
computation.  QUIC's PTO includes the peer's maximum expected acknowledgement
delay instead of using a fixed minimum timeout. Unlike TCP, which collapses
the congestion window upon expiry of an RTO, QUIC does not collapse the

> congestion window until persistent congestion {{persistent-congestion}} is

Reviewer 2: ```suggestion
congestion window until persistent congestion ({{persistent-congestion}}) is
```

declared and instead allows probe packets to temporarily exceed the congestion
window whenever the timer expires.  In practice, this is similar to TCP with

> F-RTO, but it does allow more packets to be sent when the congestion window was

Reviewer 2: Citation for F-RTO?

Author: Done with your suggestion.

> not fully utilized prior to the probe timeout expiring. Though this is slightly

Reviewer 2: ```suggestion
window whenever the timer expires.  In comparison to TCP with F-RTO
{{!RFC5682}}, which corrects an unnecessary collapse of the congestion window,
this design aims to prevent the window from collapsing. Though this is slightly
```

Perhaps?

more aggressive than TCP RTO, it's less aggressive than if the connection was

> not application limited.

Reviewer: ```suggestion
window whenever the timer expires.
```

Author: This sentence is part of the purpose of this PR from my perspective, so I think we need something along these lines.

> not application limited.

Reviewer: ```suggestion
window whenever the timer expires. In doing this, QUIC avoids congestion window reductions that
might otherwise be caused by spurious retransmissions, obviating the need for correcting
mechanisms such as F-RTO {{cite}}.

Since QUIC does not collapse the congestion window on a PTO expiration, a QUIC sender is not
limited from sending more application data after a PTO expiration. A sender could be application-limited,
have a PTO timer expire, and then send as much application data as the congestion window allows after.
At first glance this might seem to be more aggressive than TCP's RTO mechanism, where a sender
is effectively limited to sending a single packet after an RTO expiration. However, TCP's RTO period is
more accurately modeled in QUIC by the persistent congestion event, after which QUIC also limits the
sender's congestion window. 
```

Author: I took most of this, but not all of it.

A single packet loss at the tail does not indicate persistent congestion, so
QUIC specifies a time-based definition (see {{persistent-congestion}}) to
ensure one or more packets are sent prior to a dramatic decrease in

> congestion window.

Reviewer 2: ```suggestion
QUIC specifies a time-based definition to ensure one or more packets are sent
prior to a dramatic decrease in congestion window; see
{{persistent-congestion}}.
```

when the corresponding acknowledgment is sent, allowing a peer to maintain a
more accurate round-trip time estimate (see Section 13.2 of {{QUIC-TRANSPORT}}).

### Probe Timeout Replaces RTO and TLP

QUIC uses a probe timeout (see {{pto}}), with a timer based on TCP's RTO
computation.  QUIC's PTO includes the peer's maximum expected acknowledgement
delay instead of using a fixed minimum timeout. Unlike TCP, which collapses
the congestion window upon expiry of an RTO, QUIC does not collapse the
congestion window until persistent congestion ({{persistent-congestion}}) is

> declared and instead allows probe packets to temporarily exceed the congestion

Reviewer 2: ```suggestion
congestion window until persistent congestion is declared; see 
{{persistent-congestion}}. Instead allows probe packets to temporarily exceed the congestion
```

> window whenever the timer expires.

Reviewer 2: ```suggestion
congestion window until persistent congestion is declared; see
{{persistent-congestion}}. Instead QUIC allows probe packets to temporarily
exceed the congestion window whenever the timer expires.
```

when the corresponding acknowledgment is sent, allowing a peer to maintain a
more accurate round-trip time estimate (see Section 13.2 of {{QUIC-TRANSPORT}}).

### Probe Timeout Replaces RTO and TLP

> QUIC uses a probe timeout (see {{pto}}), with a timer based on TCP's RTO

Reviewer 3: RTO abbreviation is spelled out in a later chapter, but it would be helpful to state what RTO means. To a lesser extend this also applies to the section where PTO is defined because RTO is mentioned, but there is comes with a TCP doc reference and is not needed for understanding PTO.

computation.  QUIC's PTO includes the peer's maximum expected acknowledgement
delay instead of using a fixed minimum timeout. QUIC does not collapse the
congestion window until persistent congestion ({{persistent-congestion}}) is
declared, unlike TCP, which collapses the congestion window upon expiry of an
RTO.  Instead of collapsing the congestion window and declaring everything
in-flight lost, QUIC allows probe packets to temporarily exceed the congestion
window whenever the timer expires.

In doing this, QUIC avoids unnecessary congestion window reductions, obviating
the need for correcting mechanisms such as F-RTO {{!RFC5682}}. Since QUIC does
not collapse the congestion window on a PTO expiration, a QUIC sender is not
limited from sending more in-flight packets after a PTO expiration if it still

> has available congestion window. This occurs when a sender is

Reviewer 3: sounds like a word is missing before "congestion window".
